kernel/src/net/socket/netlink/route/kernel/mod.rs

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module defines the kernel socket,
+//! which is responsible for handling requests from user space.
+
+use core::marker::PhantomData;
+
+use util::{add_multi_flag, append_done_segment};
+
+use super::message::{Message, MsgSegment};
+use crate::{
+    net::socket::netlink::route::message::{CSegmentType, ErrorSegment},
+    prelude::*,
+};
+
+mod addr;
+mod link;
+mod util;
+
+pub struct NetlinkRouteKernelSocket {
+    _private: PhantomData<()>,
+}
+
+impl NetlinkRouteKernelSocket {
+    const fn new() -> Self {
+        Self {
+            _private: PhantomData,
+        }
+    }
+
+    pub fn request<F: FnMut(Message)>(&self, request: &Message, mut consume_response: F) {
+        debug!("netlink route request: {:?}", request);
+
+        for segment in request.segments() {
+            let request_header = segment.header();
+            let Ok(segment_type) = CSegmentType::try_from(request_header.type_) else {
+                // FIXME: The error is currently silently ignored.
+                // Should we respond with an error segment for this and subsequent ignored errors?
+                warn!("Invalid segment type");
+                return;
+            };
+            let response_segments = match segment {
+                MsgSegment::Link(request_segment) => {
+                    match segment_type {
```

Comment:
I still don't understand why we need to check the segment type even if we have already mapped the segment to a `MsgSegment::Link`?
```rust
        let segment = match CSegmentType::try_from(header.type_)? {
            CSegmentType::GETLINK => MsgSegment::Link(LinkSegment::read_from(header, reader)?),
            CSegmentType::GETADDR => MsgSegment::Addr(AddrSegment::read_from(header, reader)?),
            _ => return_errno_with_message!(Errno::EINVAL, "unsupported segment type"),
        };
```

Reply:
MsgSegment::Link only indicates that the request is related to a link, but it doesn't specify the operation type (e.g., add, get, delete).

We need to implement handling for these additional operations. Once these operations are supported, the error handling path can be removed.

```rust
let response_segments = match segment {
  MsgSegment::Link(request_segment) => {
     match segment_type {
        CSegmentType::GETLINK => link::do_get_link(request_segment),
        CSegmentType::ADDLINK => link::do_add_link(request_segment),
        CSegmentType::NETLINK => link::do_new_link(request_segment),
        CSegmentType::DELETELINK => link::do_delete_link(request_segment),
        _ => unreachable!()
      }
```

Reply:
OK. Thank you for your explanation. It makes sense. I originally thought of this code pattern. Could you please explain why it doesn't work?
```rust
        let segment = match CSegmentType::try_from(header.type_)? {
            CSegmentType::GETLINK => MsgSegment::GetLink(LinkSegment::read_from(header, reader)?),
            CSegmentType::ADDLINK => MsgSegment::AddLink(LinkSegment::read_from(header, reader)?),
            CSegmentType::DELETELINK => MsgSegment::DelLink(LinkSegment::read_from(header, reader)?),
            CSegmentType::GETADDR => MsgSegment::GetAddr(AddrSegment::read_from(header, reader)?),
            // ...
        };
```